src/limiters/limiters.jl

```diff
@@ -0,0 +1,52 @@
+
+"""
+    PositivityPreservingLimiterZhangShu(; threshold, variables)
+
+The fully-discrete positivity-preserving limiter of
+- Zhang, Shu (2011)
+  Maximum-principle-satisfying and positivity-preserving high-order schemes
+  for conservation laws: survey and new developments
+  [doi: 10.1098/rspa.2011.0153](https://doi.org/10.1098/rspa.2011.0153)
+The limiter is applied to all scalar `variables` in their given order
+using the associated `thresholds` to determine the minimal acceptable values.
+"""
+struct PositivityPreservingLimiterZhangShu{N, Thresholds<:NTuple{N,<:Real}, Variables<:NTuple{N,Any}}
+  thresholds::Thresholds
+  variables::Variables
+end
+
+function PositivityPreservingLimiterZhangShu(; thresholds, variables)
+  PositivityPreservingLimiterZhangShu(thresholds, variables)
+end
+
+
+function (limiter!::PositivityPreservingLimiterZhangShu)(
+    u_ode::AbstractVector, f, semi::AbstractSemidiscretization, t)
+  u = wrap_array(u_ode, semi)
+  @timeit_debug timer() "positivity-preserving limiter" limiter_zhang_shu!(
+    u, limiter!.thresholds, limiter!.variables, mesh_equations_solver_cache(semi)...)
+end
+
+
+# iterate over tuples in a type-stable way
+function limiter_zhang_shu!(u::AbstractArray{<:Any},
```

Comment:
I think you should leave a more detailed comment on the recursive implementation, e.g., how it works, what it's benefits are etc. Nothing fancy, but enough such that it can be understood in broad terms by someone who has never used (or seen) such a design pattern before.

Reply:
OTOH, is it really necessary to use such a pattern or could it be handled by a simple loop just as well? I just wonder if the zero-cost abstractions here are really required for performance.

Reply:
In this particular case for tuples of two elements, it's indeed not necessary. However, this general design pattern is type-stable, in contrast to simple approaches like https://github.com/trixi-framework/Trixi.jl/pull/186#discussion_r490709234. Thus, I think it's good to introduce preferred design patterns in Trixi such that people get familiar with them and can use them elsewhere (where it might be necessary) for performance, e.g. if the number of elements of the tuples increases.
```julia
julia> function foo(args)
           res = 0
           for i in eachindex(args)
               res += args[i](1)
           end
           res
       end
foo (generic function with 1 method)

julia> foo((+, -, sin, cos))
1.3817732906760363

julia> @code_warntype foo((+, -, sin, cos))
Variables
  #self#::Core.Compiler.Const(foo, false)
  args::Core.Compiler.Const((+, -, sin, cos), false)
  res::Any
  @_4::Union{Nothing, Tuple{Int64,Int64}}
  i::Int64

Body::Any
1 ─       (res = 0)
│   %2  = Main.eachindex(args)::Core.Compiler.Const(Base.OneTo(4), false)
│         (@_4 = Base.iterate(%2))
│   %4  = (@_4::Core.Compiler.Const((1, 1), false) === nothing)::Core.Compiler.Const(false, false)
│   %5  = Base.not_int(%4)::Core.Compiler.Const(true, false)
└──       goto #4 if not %5
2 ┄ %7  = @_4::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│         (i = Core.getfield(%7, 1))
│   %9  = Core.getfield(%7, 2)::Int64
│   %10 = res::Any
│   %11 = Base.getindex(args, i)::Any
│   %12 = (%11)(1)::Any
│         (res = %10 + %12)
│         (@_4 = Base.iterate(%2, %9))
│   %15 = (@_4 === nothing)::Bool
│   %16 = Base.not_int(%15)::Bool
└──       goto #4 if not %16
3 ─       goto #2
4 ┄       return res

julia> function bar(args::NTuple{N,Any}) where {N}
           arg1 = first(args)
           remaining_args = Base.tail(args)
           arg1(1) + bar(remaining_args)
       end
bar (generic function with 1 method)

julia> bar(::Tuple{}) = 0
bar (generic function with 2 methods)

julia> bar((+, -, sin, cos))
1.3817732906760363

julia> @code_warntype bar((+, -, sin, cos))
Variables
  #self#::Core.Compiler.Const(bar, false)
  args::Core.Compiler.Const((+, -, sin, cos), false)
  arg1::typeof(+)
  remaining_args::Tuple{typeof(-),typeof(sin),typeof(cos)}

Body::Float64
1 ─      (arg1 = Main.first(args))
│   %2 = Base.tail::Core.Compiler.Const(Base.tail, false)
│        (remaining_args = (%2)(args))
│   %4 = (arg1)(1)::Core.Compiler.Const(1, false)
│   %5 = Main.bar(remaining_args)::Float64
│   %6 = (%4 + %5)::Float64
└──      return %6
```

Reply:
I've added additional comments in c75b6bd.

Reply:
Thanks! 👍